### PR TITLE
Restart service automatically

### PR DESCRIPTION
**Pull request by hiroxy**
Added the ability to automatically restart Silver on failure. At this stage, I've hardcoded 10 secs as a delay before restarting in case of failure, but should it be configurable?
I will change the draft pull request to a formal one once we have decided whether to keep the hardcoded delay or not, so please look over the code and point out any changes that need to be made.

### CODE REVIEW

File: service/svcutil/svcutil_windows.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	"time"
+
+	"golang.org/x/sys/windows/svc/mgr"
+)
+
+// SetServiceToRestart sets the service named name to automatically restart after waiting for a duration specified by
+// delay when the service fails, and the time after which to reset the service failure count to zero if there are no
+// failures, in seconds. Specify systemcall.INFINITE to indicate that service failure count should never be reset. The
+// service has to already be registered in the SCM.
+func SetServiceToRestart(name string, delay time.Duration, resetPeriod uint32) error {
```
Comment (hiroxy) [inline service/svcutil/svcutil_windows.go:17]: Probably `resetPeriod` should be `time.Duration`, and set INFINITY internally if the value is negative.

File: service/svcutil/svcutil_windows.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+const (
+	MaxDelay            = syscall.INFINITE * time.Millisecond
+	InfiniteResetPeriod = syscall.INFINITE * time.Second
+)
+
+// SetServiceToRestart sets the service named name to automatically restart after waiting for a duration specified by
+// delay when the service fails, and the time after which to reset the service failure count to zero if there are no
+// failures. The service that specifies has to already be registered in the SCM. If delay is negative or greater than
+// MaxDelay, or resetPeriod is negative, an error is returned. If resetPeriod is grater than or equal
```
Comment (amir-khassaia) [inline service/svcutil/svcutil_windows.go:23]: grater => greater

File: service/svcutil/svcutil.go [hunk 3]
```diff
@@ -29,6 +29,13 @@ func init() {
 	random = rand.New(rand.NewSource(time.Now().UTC().UnixNano() + int64(os.Getpid())))
 }
 
+// RestartConfig is a configuration for restarting a service.
+type RestartConfig struct {
+	ServiceName  string        // Service name to be configured. It should be already registered in SCM.
+	RestartDelay time.Duration // Time to wait before rebooting
```
[the remark this answers is not in the document]
Comment (hiroxy) [reply service/svcutil/svcutil.go:35]: yes, "restart" is better. will change.

File: service/main.go [hunk 4]
```diff
@@ -40,57 +40,76 @@ type context struct {
...
+		if err = writeProxyConf(); err != nil {
 			fmt.Fprintf(os.Stderr, "WARNING: Unable to store HTTP Proxy settings: %v\n", err)
 		}
+		defer func() {
+			if exitCode != 0 {
+				return
+			}
+			restartConfig := svcutil.RestartConfig{
+				ServiceName:  serviceName(),
+				RestartDelay: 10 * time.Second,
```
[the remark this answers is not in the document]
Comment (hiroxy) [reply service/main.go:86]: If there are needs for it, we can do so. But I think that it's probably not necessary at least at this stage.

File: service/svcutil/svcutil.go [hunk 5]
```diff
@@ -29,6 +29,13 @@ func init() {
...
+	ResetPeriod  time.Duration // Time after which to rest the service failure count to zero if there are no failures
```
[the remark this answers is not in the document]
Comment (hiroxy) [reply service/svcutil/svcutil.go:36]: will use `ResetFailCountAfter`

File: service/svcutil/svcutil_windows.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+// InfiniteResetPeriod, the service failure count is never reset.
+func SetServiceToRestart(conf RestartConfig) error {
+	switch {
+	case conf.RestartDelay < 0:
+		return errors.New("Invalid delay time")
```
[the remark this answers is not in the document]
Comment (hiroxy) [reply service/svcutil/svcutil_windows.go:28]: I thought so, but since all the other places use capital letters, I followed it.

**Review by amir-khassaia — APPROVED** [review]

**Pull request merged**